engine/execution/storehouse/register_store.go

```diff
@@ -0,0 +1,266 @@
+}
+
+// getAndConvertNotFoundErr returns nil if the register is not found from storage
+func (r *RegisterStore) getAndConvertNotFoundErr(register flow.RegisterID, height uint64) (flow.RegisterValue, error) {
+	val, err := r.diskStore.Get(register, height)
+	if errors.Is(err, storage.ErrNotFound) {
+		// FVM expects the error to be nil when register is not found
+		return nil, nil
+	}
+	return val, err
+}
+
+// SaveRegisters saves to InMemoryRegisterStore first, then trigger the same check as OnBlockFinalized
+// Depend on InMemoryRegisterStore.SaveRegisters
+// It returns:
+// - nil if the registers are saved successfully
+// - exception is the block is above the pruned height but does not connect to the pruned height (conflicting block).
+// - exception if the block is below the pruned height
+// - exception if the save block is saved again
+// - exception for any other exception
+func (r *RegisterStore) SaveRegisters(header *flow.Header, registers flow.RegisterEntries) error {
+	err := r.memStore.SaveRegisters(header.Height, header.ID(), header.ParentID, registers)
+	if err != nil {
+		return fmt.Errorf("cannot save register to memStore: %w", err)
+	}
+
+	err = r.OnBlockFinalized()
+	if err != nil {
+		return fmt.Errorf("cannot trigger OnBlockFinalized: %w", err)
+	}
+	return nil
+}
+
+// Depend on FinalizedReader's FinalizedBlockIDAtHeight
+// Depend on ExecutedFinalizedWAL.Append
+// Depend on OnDiskRegisterStore.SaveRegisters
+// OnBlockFinalized trigger the check of whether a block at the next height becomes finalized and executed.
+// the next height is the existing finalized and executed block's height + 1.
+// If a block at next height becomes finalized and executed, then:
+// 1. write the registers to write ahead logs
+// 2. save the registers of the block to OnDiskRegisterStore
+// 3. prune the height in InMemoryRegisterStore
+func (r *RegisterStore) OnBlockFinalized() error {
+	// only one goroutine can execute OnBlockFinalized at a time
+	if !r.finalizing.CompareAndSwap(false, true) {
+		return nil
+	}
+
+	defer r.finalizing.Store(false)
+	return r.onBlockFinalized()
+}
+
+func (r *RegisterStore) onBlockFinalized() error {
+	latest := r.diskStore.LatestHeight()
+	next := latest + 1
+	blockID, err := r.finalized.FinalizedBlockIDAtHeight(next)
+	if errors.Is(err, storage.ErrNotFound) {
+		// next block is not finalized yet
+		return nil
+	}
+
+	regs, err := r.memStore.GetUpdatedRegisters(next, blockID)
+	if errors.Is(err, ErrNotExecuted) {
+		// next block is not executed yet
+		return nil
+	}
+
+	// TODO: append WAL
+	// err = r.wal.Append(next, regs)
+	// if err != nil {
+	// 	return fmt.Errorf("cannot write %v registers to write ahead logs for height %v: %w", len(regs), next, err)
+	// }
+
+	err = r.diskStore.Store(regs, next)
+	if err != nil {
+		return fmt.Errorf("cannot save %v registers to disk store for height %v: %w", len(regs), next, err)
+	}
+
+	err = r.memStore.Prune(next, blockID)
```

Comment:
is it ok to prune with every block finalized, would it be a valuable improvement if we only prune once we don't have more finalized blocks? not sure how expensive is to prune just an idea.

Reply:
prune should be fast and cheap, since it's just acquire a lock and updating some index in memory. 

I think the performance is quite similar comparing the following two approach:
1. pruning once to prune 3 finalized blocks
2. pruning 3 times each time to prune 1 finalized block